File: lib/features/saved_billers/data/saved_biller.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'saved_biller.freezed.dart';

@freezed
abstract class SavedBiller with _$SavedBiller {
  const factory SavedBiller({
    required String billerId,
    required String account,
    required String nickname,
  }) = _SavedBiller;
}

@freezed
abstract class SavedBillers with _$SavedBillers {
  const SavedBillers._();

  const factory SavedBillers({required List<SavedBiller> items}) =
      _SavedBillers;

  factory SavedBillers.empty() => const SavedBillers(items: []);

  bool contains(String billerId, String account) => items
      .any((s) => s.billerId == billerId && s.account == account);

  SavedBillers adding(SavedBiller saved) =>
      SavedBillers(items: [...items, saved]);

  SavedBillers removing(String billerId, String account) => SavedBillers(
      items: items
          .where((s) => s.billerId != billerId || s.account != account)
          .toList());
}

File: src/features/saved_billers/data/savedBiller.ts
export type SavedBiller = { billerId: string; account: string; nickname: string };

export type SavedBillers = { items: SavedBiller[] };

export const emptySavedBillers = (): SavedBillers => ({ items: [] });

export const savedContains = (saved: SavedBillers, billerId: string, account: string): boolean =>
  saved.items.some((s) => s.billerId === billerId && s.account === account);

export const addingSaved = (saved: SavedBillers, entry: SavedBiller): SavedBillers => ({
  items: [...saved.items, entry],
});

export const removingSaved = (saved: SavedBillers, billerId: string, account: string): SavedBillers => ({
  items: saved.items.filter((s) => s.billerId !== billerId || s.account !== account),
});
